pkg/cloudprovider/provider/anexia/provider.go

```diff
@@ -0,0 +1,346 @@
+		return errors.New("location id is missing")
+	}
+
+	if config.TemplateID == "" {
+		return errors.New("template id is missing")
+	}
+
+	if config.VlanID == "" {
+		return errors.New("vlan id is missing")
+	}
+
+	return nil
+}
+
+func (p *provider) Get(machine *v1alpha1.Machine, _ *cloudprovidertypes.ProviderData) (instance.Instance, error) {
+	config, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, newError(common.InvalidConfigurationMachineError, "failed to parse MachineSpec: %v", err)
+	}
+
+	apiClient := getClient(config.Token)
+	status, err := getStatus(machine.Status.ProviderStatus)
+	if err != nil {
+		return nil, newError(common.InvalidConfigurationMachineError, "failed to get machine status: %v", err)
+	}
+	if status.InstanceID == "" {
+		return nil, cloudprovidererrors.ErrInstanceNotFound
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), anxtypes.GetRequestTimeout)
+	defer cancel()
+
+	info, err := apiClient.VSphere().Info().Get(ctx, status.InstanceID)
+	if err != nil {
+		return nil, fmt.Errorf("failed get machine info: %w", err)
+	}
+
+	return &anexiaInstance{
+		info: &info,
+	}, nil
+}
+
+func (p *provider) GetCloudConfig(spec v1alpha1.MachineSpec) (string, string, error) {
+	return "", "", nil
+}
+
+// Create creates a cloud instance according to the given machine
+func (p *provider) Create(machine *v1alpha1.Machine, providerData *cloudprovidertypes.ProviderData, userdata string) (instance.Instance, error) {
+	config, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, newError(common.InvalidConfigurationMachineError, "failed to parse MachineSpec: %v", err)
+	}
+
+	apiClient := getClient(config.Token)
+	ctx, cancel := context.WithTimeout(context.Background(), anxtypes.CreateRequestTimeout)
+	defer cancel()
+
+	status, err := getStatus(machine.Status.ProviderStatus)
+	if err != nil {
+		return nil, newError(common.InvalidConfigurationMachineError, "failed to get machine status: %v", err)
+	}
+
+	if status.ProvisioningID == "" {
+		ips, err := apiClient.VSphere().Provisioning().IPs().GetFree(ctx, config.LocationID, config.VlanID)
+		if err != nil {
+			return nil, newError(common.InvalidConfigurationMachineError, "failed to get ip pool: %v", err)
+		}
+		if len(ips) < 1 {
+			return nil, newError(common.InsufficientResourcesMachineError, "no ip address is available for this machine")
+		}
+
+		ipID := ips[0].Identifier
+		networkInterfaces := []anxvm.Network{{
+			NICType: anxtypes.VmxNet3NIC,
+			IPs:     []string{ipID},
+			VLAN:    config.VlanID,
+		}}
+
+		vm := apiClient.VSphere().Provisioning().VM().NewDefinition(
+			config.LocationID,
+			"templates",
+			config.TemplateID,
+			machine.ObjectMeta.Name,
+			config.CPUs,
+			config.Memory,
+			config.DiskSize,
+			networkInterfaces,
+		)
+
+		vm.Script = base64.StdEncoding.EncodeToString(
+			[]byte(fmt.Sprintf("anexia: true\n\n%s", userdata)),
+		)
+
+		sshKey, err := ssh.NewKey()
```

Comment:
Why this is needed? It is not possible to create a vm without introducing an ssh key?

Reply:
@erdii could you elaborate? As I understood from some conversation some controller is adding later our ssh keys but it needs to have generated key-pair first to use it?

Reply:
Looking at the API I think you must add either an ssh key or a password, so I guess yeah this is needed.